First DOM Exercise, validating a form

augusticor opened the pull request:
Validating the form in HTML-Tutorial using JavaScript, document object model

augusticor committed:
First DOM Exercise, validating a form

juancho11gm commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,70 @@
+const EMAIL_V = /([a-z0-9]+[_az0-9\.-]*[a-z0-9]+)@([a-z0-9-]+(?:\.[a-z0-9-]+)*\.[a-z]{2,4})/;
+const PHONE_V = /[0-9]{2,}/;
+
+var htmlForm = document.querySelector('#form');
+htmlForm.addEventListener('submit', validateForm);
+var finalMessages = document.querySelector('#final-information');
+
+function validateForm(event) {
+    event.preventDefault();
+    let form = {
+        name: document.querySelector('#name'),
+        email: document.querySelector('#email'),
+        phone: document.querySelector('#phonen'),
+        courses: document.querySelectorAll('input[type="checkbox"]'),
+        message: document.querySelector('#texta')
+    };
+    printInHTML(form.name, form.email, form.phone, form.courses, form.message);
+}
```
You can do something like this. The inputs will be available to other functions.
```
const htmlForm = document.querySelector('#form');
const courses = document.querySelectorAll('input[type="checkbox"]');
const name = document.querySelector('#name');
const email = document.querySelector('#email');
const phone =   document.querySelector('#phonen');
const message = document.querySelector('#texta');

let form = {
    name:    name,
    email:   email,
    phone:   phone,
    courses: courses,
    message: message
};

let finalMessages = document.querySelector('#final-information');
htmlForm.addEventListener('submit', validateForm);

function validateForm(event) {
    event.preventDefault();
    printInHTML(form);
}

juancho11gm commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,70 @@
+    printInHTML(form.name, form.email, form.phone, form.courses, form.message);
+}
+
+function validateName(nameToValidate) {
+    if (nameToValidate.value == '') {
+        nameToValidate.classList.add('name-error');
+        return 'Please fill the name field !';
+    }
+    else if (!NAME_V.test(nameToValidate.value)) {
+        nameToValidate.classList.add('name-error');
+        return 'Please write a valid name, not numbers or special characters !';
+    }
+    nameToValidate.classList.remove('name-error');
+    return nameToValidate.value;
+}
```
you can change the name of the parameter to `name`. With the name of the function, we already know that it's validating.

juancho11gm commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,70 @@
+    return nameToValidate.value;
+}
+
+function validateEmail(emailToValidate) {
+    if(emailToValidate.value == '') {
+        emailToValidate.classList.add('email-error');
+        return 'Please fill out the email space !';
+    }
+    else if (!EMAIL_V.test(emailToValidate.value)) {
+        emailToValidate.classList.add('email-error');
+        return 'Please write a valid email';
+    }
+    emailToValidate.classList.remove('email-error');
+    return emailToValidate.value;
+}
```
you can change the name of the parameter to `email`. With the name of the function, we already know that it's validating.

juancho11gm commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,70 @@
+}
+
+function validatePhoneNumber(phoneToValidate) {
+    return PHONE_V.test(phoneToValidate) ? phoneToValidate : 'Please write at least 2 numbers !';
```
you can change the name of the parameter to `phone`. With the name of the function, we already know that it's validating. 
I think that you are verifying all the element, we need to check the `phone.value`

juancho11gm commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,70 @@
+    return (checked === undefined || checked.length == 0) ? 'Please select at least one course, you are interested in !' : checked;
+}
+
+function printInHTML(nameToValidate, emailToValidate, phoneToValidate, courses, message) {
```
Here you can use **destructuring**.
```
function printInHTML({name, email, phone, courses, message}) {
```

juancho11gm commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,70 @@
+}
+
+function printInHTML(nameToValidate, emailToValidate, phoneToValidate, courses, message) {
+    finalMessages.innerHTML = 'VALIDATION RESULT : <br><br>';
+    finalMessages.innerHTML += validateName(nameToValidate) + '<br>';
+    finalMessages.innerHTML += validateEmail(emailToValidate) + '<br>';
+    finalMessages.innerHTML += validatePhoneNumber(phoneToValidate) + '<br>';
+    finalMessages.innerHTML += validateCourses(courses) + '<br>';
+    finalMessages.innerHTML += message.value;
```
Print the form values in the HTML (below the form) **if all of them are correct only,**.

juancho11gm replied:
And you can create other function to render the response to the HTML with `createElement('p');` and `appendChild(element)` methods, instead of dealing with += and strings.
*Update: the main purpose of creating this function is to reuse the code.

juancho11gm commented on 2-dom-manipulation/form-styles.css:
```diff
@@ -0,0 +1,77 @@
+    font-family: 'Trebuchet MS';
+    font-size: 20px;
+    height: 60px;
+    margin-left: 30px;
+    margin-right: 30px;
+    margin-top: 10%;
```
```suggestion
    margin: 10% 30px 0 30px;
```

juancho11gm commented on 2-dom-manipulation/form-styles.css:
```diff
@@ -0,0 +1,77 @@
+}
+
+/*ERRORS*/
+.name-error {
+    border: 2px solid red;
+}
+
+.email-error {
+    border: 2px solid red;
+}
```
```suggestion
.email-error, .name-error {
    border: 2px solid red;
}
```

juancho11gm reviewed (changes requested):
Hi Óscar, nice approach. Please see the comments and send the changes. If you have doubts let us know.

augusticor committed:
Creating extra functions and change the error messages in html

juancho11gm commented on 2-dom-manipulation/form.js:
```diff
@@ -17,37 +16,51 @@ function validateForm(event) {
     }
-    nameToValidate.classList.remove('name-error');
-    return nameToValidate.value;
+    name.classList.remove('name-error');
```
Here you are removing the class even if it's bad.

augusticor replied:
How can i specify to change the styles, "remove the error red border" ?

juancho11gm commented on 2-dom-manipulation/form.js:
```diff
@@ -17,37 +16,51 @@ function validateForm(event) {
     }
-    emailToValidate.classList.remove('email-error');
-    return emailToValidate.value;
+    email.classList.remove('email-error');
```
Here you are removing the class even if it's bad.

juancho11gm commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,70 @@
+"use strict";
```
```suggestion
'use strict';
```

juancho11gm commented on 2-dom-manipulation/form.js:
```diff
@@ -57,14 +70,21 @@ function validateCourses(courses) {
         }
     }
-    return (checked === undefined || checked.length == 0) ? 'Please select at least one course, you are interested in !' : checked;
+    if (checked === undefined || checked.length == 0) {
```
```suggestion
    if (!checked.length) {
```

juancho11gm commented on 2-dom-manipulation/form.js:
```diff
@@ -17,37 +16,51 @@ function validateForm(event) {
+    let name = validateName(form.name);
+    let email = validateEmail(form.email);
+    let phone = validatePhoneNumber(form.phone);
+    let courses = Array.isArray(validateCourses(form.courses));
+    if (name && email && phone && courses) {
+        printInHTML(form.name.value);
+        printInHTML(form.email.value);
+        printInHTML(form.phone.value);
+        printInHTML(validateCourses(form.courses));
```
it's necessary to call validateCourses twice? (please let me know).

juancho11gm replied:
Also you can use `for…in` to iterate over the properties of an object (the object keys):

augusticor replied:
I had to call the function twice because if I call the functions into the if statement, it trunks if for example no course is selected, and all the other elements are correct.

juancho11gm commented:
Print the form values in the HTML (below the form) if **all of them are correct only**. Also you can put outside the form object.

augusticor commented:
@juancho11gm 
In the JavaScript tutorial website says  

>  Print the form values in the HTML (below the form) if all of them are correct only, **display the corresponding message if the form has an error**.

Where do I display that form error ?

juancho11gm commented:
> @juancho11gm
> In the JavaScript tutorial website says
> 
> > Print the form values in the HTML (below the form) if all of them are correct only, **display the corresponding message if the form has an error**.
> 
> Where do I display that form error ?

I mean that if you try with a wrong name-email-phone and select at least one course it will display the courses. Also the .name-error .email-error classes are not being applied.

augusticor committed:
Correcting display in HTML errors

augusticor committed:
Validating validations functionallity

pushdev-code commented on 2-dom-manipulation/form-styles.css:
```diff
@@ -0,0 +1,76 @@
+    text-align: center;
+}
+
+button {
```
prefer classnames over tagnames.

pushdev-code commented on 2-dom-manipulation/form-styles.css:
```diff
@@ -0,0 +1,76 @@
+    width: 200px;
+}
+
+form {
```
prefer classnames over tagnames.

pushdev-code commented on 2-dom-manipulation/form-styles.css:
```diff
@@ -0,0 +1,76 @@
+    padding: 40px;
+}
+
+fieldset {
```
prefer classnames over tagnames.

pushdev-code commented on 2-dom-manipulation/form-styles.css:
```diff
@@ -0,0 +1,76 @@
+
+#send:hover {
+    background-color: lightslategray;
+    color: white;
```
this is redundant, it's being applied in line 59.

pushdev-code commented on 2-dom-manipulation/form-styles.css:
```diff
@@ -0,0 +1,76 @@
+}
+
+/*ERRORS*/
+.email-error, .name-error {
```
```suggestion
.email-error, 
.name-error {
```

pushdev-code commented on 2-dom-manipulation/form-styles.css:
```diff
@@ -0,0 +1,76 @@
+    padding-top: 30px;
+}
+
+p {
```
prefer classnames over tagnames.

pushdev-code commented on 2-dom-manipulation/form-styles.css:
```diff
@@ -0,0 +1,76 @@
+    font-family: 'Trebuchet MS';
+}
+
+#reset {
```
don't use ids, use class selectors.

augusticor replied:
Ok, ids just for JavaScript ?, I don't have any other button with same name or action ...

augusticor replied:
And I can't do that because it now has 'buttons' class

pushdev-code replied:
you can add an extra class to the button. `<button class="button js-button"></button>` and use `.js-button` inside Javascript.

pushdev-code commented on 2-dom-manipulation/form.html:
```diff
@@ -0,0 +1,62 @@
+                </label>
+
+                <label for="phonen">
+                    Phone number (+57): <input type="number" name="phonenumber" id="phonen" min="1" aria-label="your phone number +57 extension" />
```
```suggestion
                    Phone number (+57): <input type="tel" name="phonenumber" id="phonen" min="1" aria-label="your phone number +57 extension" />
```

pushdev-code commented on 2-dom-manipulation/form.html:
```diff
@@ -0,0 +1,62 @@
+            </fieldset>
+
+            <fieldset class="field-courses">
+                <legend>Please select the courses that you are interested in <abbr title="required" aria-label="The courses you are interested to take">*</abbr></legend>
```
```suggestion
                <legend id="courses-legend">Please select the courses that you are interested in <abbr title="required" aria-labelledby="courses-legend">*</abbr></legend>
```

augusticor replied:
This aria label takes the text from the legend ?

pushdev-code commented on 2-dom-manipulation/form.html:
```diff
@@ -0,0 +1,62 @@
+                </label>
+            </fieldset>
+
+            <p>Please enter a message</p>
```
```suggestion
            <label for="texta">Please enter a message</label>
```

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,91 @@
+'use strict';
+
+const NAME_V = /^(([A-Za-z\s])+)$/;
```
```suggestion
const NAME_REGEX = /^(([A-Za-z\s])+)$/;
```

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,91 @@
+'use strict';
+
+const NAME_V = /^(([A-Za-z\s])+)$/;
+const EMAIL_V = /([a-z0-9]+[_az0-9\.-]*[a-z0-9]+)@([a-z0-9-]+(?:\.[a-z0-9-]+)*\.[a-z]{2,4})/;
```
```suggestion
const EMAIL_REGEX = /([a-z0-9]+[_az0-9\.-]*[a-z0-9]+)@([a-z0-9-]+(?:\.[a-z0-9-]+)*\.[a-z]{2,4})/;
```

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,91 @@
+        email: document.querySelector('#email'),
+        phone: document.querySelector('#phonen'),
+        courses: document.querySelectorAll('input[type="checkbox"]'),
+        message: document.querySelector('#texta')
```
```suggestion
        name: HTML_FORM.querySelector('#name'),
        email: HTML_FORM.querySelector('#email'),
        phone: HTML_FORM.querySelector('#phonen'),
        courses: HTML_FORM.querySelectorAll('input[type="checkbox"]'),
        message: HTML_FORM.querySelector('#texta')
```

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,91 @@
+    }
+    if (!checked.length) {
+        alert('Please select at least one course, you are interested in');
+        document.querySelector('.field-courses legend').classList.add('fieldset-courses-legend-error');
```
save `document.querySelector('.field-courses legend')` in a const so it can be reused, here you're forcing to make an unnecessary query to the DOM every time the form is submitted.

pushdev-code replied:
save `fieldset-courses-legend-error` in a const and reuse it.

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,91 @@
+
+function validateName(name) {
+    if (name.value == '') {
+        name.classList.add('name-error');
```
save `name-error` in a const and reuse it.

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,91 @@
+        return false;
+    }
+    else if (!NAME_V.test(name.value)) {
+        name.classList.add('name-error');
```
save `name-error` in a const and reuse it.

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,91 @@
+        return false;
+    }
+    else {
+        form.name.classList.remove('name-error');
```
save `name-error` in a const and reuse it.

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,91 @@
+
+function validateEmail(email) {
+    if (email.value == '') {
+        email.classList.add('email-error');
```
save `email-error` in a const and reuse it.

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,91 @@
+        return false;
+    }
+    else if (!EMAIL_V.test(email.value)) {
+        email.classList.add('email-error');
```
save `email-error` in a const and reuse it.

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,91 @@
+        return false;
+    }
+    else {
+        form.email.classList.remove('email-error');
```
save `email-error` in a const and reuse it.

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,91 @@
+        document.querySelector('.field-courses legend').classList.add('fieldset-courses-legend-error');
+        return '';
+    } else {
+        document.querySelector('.field-courses legend').classList.remove('fieldset-courses-legend-error');
```
save `fieldset-courses-legend-error` in a const and reuse it.

augusticor committed:
Creating extra constants and changing html tags to classes for styling

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,105 @@
+function validateName(name) {
+    if (name.value == '') {
+        name.classList.add(NAME_ERROR_CLASS);
+        alert('Please fill the name field !');
```
avoid alerts, they're intrusive and blocking.

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,105 @@
+const HTML_FORM = document.querySelector('.form');
+HTML_FORM.addEventListener('submit', validateForm);
+
+const NAME_ERROR_CLASS = 'name-error';
+const EMAIL_ERROR_CLASS = 'email-error';
+const PHONE_ERROR_CLASS = 'phone-error';
+const COURSES_ERROR_CLASS = 'fieldset-courses-legend-error';
+const FIELD_COURSES_LEGEND = document.querySelector('.field-courses legend');
```
```suggestion
const CLASSES = {
NAME_ERROR_CLASS : 'name-error',
EMAIL_ERROR_CLASS : 'email-error',
PHONE_ERROR_CLASS : 'phone-error',
COURSES_ERROR_CLASS : 'fieldset-courses-legend-error',
FIELD_COURSES_LEGEND : document.querySelector('.field-courses legend')
};
```
You can use enums or objects in JS and use `CLASSES. NAME_ERROR_CLASS` where you need the value.

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,105 @@
+}
+
+function printInHTML(text) {
+    let p = document.createElement('p');
```
you don't need to create an element everytime there's an error, what you can do is to have these elements already on the markup and keep them non visible with `display:none;` when you need to display the message you just need to add/remove the class so they're visible without creating new elements.

pushdev-code replied:
Check this example I've created: http://plnkr.co/edit/OUlHbkmxcRA1J3IF

augusticor committed:
Update the presentation of error and final messages

- Changed the class for button
- Updated the presentation from alerts to p messages when the user
writes wrong data
- Updated the elements for not creating many times

pushdev-code commented on 2-dom-manipulation/form.js:
```diff
@@ -0,0 +1,116 @@
+    phone: html_form.querySelector('#phonen'),
+    courses: html_form.querySelectorAll('input[type="checkbox"]'),
+    message: html_form.querySelector('#texta')
+}
```
```suggestion
};
```

augusticor committed:
Update 2-dom-manipulation/form.js

; line 17

Co-authored-by: Push Dev Code <61557537+pushdev-code@users.noreply.github.com>
